File: backend/app/schemas/assessment.py
from pydantic import BaseModel, ConfigDict
from typing import List, Optional, Any
from datetime import datetime
from .skill import SkillResponse

class AssessmentQuestionBase(BaseModel):
    question_text: str
    options: Any
    difficulty: int

class AssessmentQuestionResponse(AssessmentQuestionBase):
    id: int
    model_config = ConfigDict(from_attributes=True)

class AssessmentBase(BaseModel):
    title: str
    description: Optional[str]
    skill_id: int

class AssessmentResponse(AssessmentBase):
    id: int
    skill: SkillResponse
    target_company_id: Optional[int] = None
    assessment_source: str = "role_fallback" # "company_specific" or "role_fallback"
    target_company_name: Optional[str] = None
    model_config = ConfigDict(from_attributes=True)


class AssessmentDetailResponse(AssessmentResponse):
    questions: List[AssessmentQuestionResponse]
    model_config = ConfigDict(from_attributes=True)

class AssessmentAnswerCreate(BaseModel):
    question_id: int
    selected_option_index: int

class AssessmentSubmission(BaseModel):
    answers: List[AssessmentAnswerCreate]

class AssessmentResultResponse(BaseModel):
    id: int
    assessment_id: int
    score: Optional[float] = None
    started_at: datetime
    completed_at: Optional[datetime] = None
    model_config = ConfigDict(from_attributes=True)

File: frontend/src/services/assessment.ts
import apiClient from './apiClient';
import { type Skill } from './skill';

export interface AssessmentQuestion {
  id: number;
  question_text: string;
  options: string[];
  difficulty: number;
}

export interface Assessment {
  id: number;
  title: string;
  description?: string;
  skill_id: number;
  skill: Skill;
  target_company_id?: number;
  assessment_source?: 'company_specific' | 'role_fallback' | string;
  target_company_name?: string;
}

export interface AssessmentDetail extends Assessment {
  questions: AssessmentQuestion[];
}

export interface AssessmentAnswer {
  question_id: number;
  selected_option_index: number;
}

export interface AssessmentSubmission {
  answers: AssessmentAnswer[];
}

export interface AssessmentResult {
  id: number;
  assessment_id: number;
  score: number;
  started_at: string;
  completed_at: string;
}

export const getAssessments = async () => {
  const { data } = await apiClient.get<Assessment[]>('/api/v1/assessments');
  return data;
};

export const getAssessment = async (id: number) => {
  const { data } = await apiClient.get<AssessmentDetail>(`/api/v1/assessments/${id}`);
  return data;
};

export const startAttempt = async (assessmentId: number) => {
  const { data } = await apiClient.post<AssessmentResult>(`/api/v1/assessments/${assessmentId}/attempt`);
  return data;
};

export const submitAttempt = async (attemptId: number, submission: AssessmentSubmission) => {
  const { data } = await apiClient.post<AssessmentResult>(`/api/v1/assessments/attempt/${attemptId}/submit`, submission);
  return data;
};
